File: python-mcp-workspace/filesystem_mcp_server/path_utils.py
from pathlib import Path

from .config import REPOSITORIES_ROOT


def validate_relative_path(relative_path: str) -> None:
    if not isinstance(relative_path, str) or not relative_path.strip():
        raise ValueError("relative_path must be a non-empty string")


def validate_content(content: str) -> None:
    if not isinstance(content, str):
        raise ValueError("content must be a string")


def resolve_inside_root(relative_path: str = ".") -> Path:
    if not isinstance(relative_path, str):
        raise ValueError("relative_path must be a string")

    resolved_path = (REPOSITORIES_ROOT / relative_path).resolve()

    try:
      resolved_path.relative_to(REPOSITORIES_ROOT)
    except ValueError as exc:
        raise ValueError("Path is outside the configured repositories root") from exc

    return resolved_path

File: filesystem-mcp-server/src/pathUtils.js
import { constants } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { repositoriesRoot } from "./config.js";

export const validateRelativePath = (relativePath) => {
  if (typeof relativePath !== "string" || relativePath.trim() === "") {
    throw new Error("relativePath must be a non-empty string");
  }
};

export const validateContent = (content) => {
  if (typeof content !== "string") {
    throw new Error("content must be a string");
  }
};

export const resolveInsideRoot = (relativePath = ".") => {
  if (typeof relativePath !== "string") {
    throw new Error("relativePath must be a string");
  }

  const resolvedPath = path.resolve(repositoriesRoot, relativePath);
  const relativeToRoot = path.relative(repositoriesRoot, resolvedPath);

  if (
    relativeToRoot === ".." ||
    relativeToRoot.startsWith(`..${path.sep}`) ||
    path.isAbsolute(relativeToRoot)
  ) {
    throw new Error("Path is outside the configured repositories root");
  }

  return resolvedPath;
};

export const pathExists = async (filePath) => {
  try {
    await fs.access(filePath, constants.F_OK);
    return true;
  } catch {
    return false;
  }
};
